disk_layout/0001-Passing-disk_profile-as-part-of-user_data.patch

```diff
@@ -0,0 +1,487 @@
++                if disk_config_data:
++                    # validate disk_profile section from user_data against
++                    # json schema
++                    disk_config = {"disk_config": disk_config_data}
++                    jsonschema.validate(disk_config, DISK_CONFIG_SCHEMA)
++                    patch.append({'path': '/instance_info/disk_config',
++                                  'op': 'add',
++                                  'value': jsonutils.dumps(disk_config_data)})
++
++            except yaml.YAMLError as e:
++                msg = "user_data is not a YAML file: %s" % e.message
++                LOG.error(msg)
++                raise exception.InstanceDeployFailure(msg)
++            except jsonschema.ValidationError as e:
++                msg = "Disk_profile schema is invalid: %s" % e.message
++                LOG.error(msg)
++                raise exception.InstanceDeployFailure(msg)
++
+         return patch
+diff --git a/nova/virt/ironic/validation.py b/nova/virt/ironic/validation.py
+new file mode 100644
+index 0000000000..350a85596f
+--- /dev/null
++++ b/nova/virt/ironic/validation.py
+@@ -0,0 +1,209 @@
++# Copyright 2017 Intel, Inc.
++# All Rights Reserved.
++#
++#    Licensed under the Apache License, Version 2.0 (the "License"); you may
++#    not use this file except in compliance with the License. You may obtain
++#    a copy of the License at
++#
++#         http://www.apache.org/licenses/LICENSE-2.0
++#
++#    Unless required by applicable law or agreed to in writing, software
++#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
++#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
++#    License for the specific language governing permissions and limitations
++#    under the License.
++
++DISK_CONFIG_SCHEMA = {
++  "$schema": "http://json-schema.org/schema#",
++  "definitions": {
++    "partition": {
++      "type": "object",
++      "properties": {
++        "size": {
++          "type": "string",
++          "description": "Absolute size of partition"
++        },
++        "minsize": {
++          "type": "string",
++          "description": "Minimum size of partition,"
++                    "if any space left expand this partition"
++        },
++        "type": {
++          "type": "string",
++          "description": "Type of partition"
++        }
++      },
++      "oneOf": [
++        {"required": ["minsize"]},
++        {"required": ["size"]}
++      ],
++      "additionalProperties": False
++    },
++    "blockdevice": {
++      "type": "object",
++      "properties": {
++        "candidates": {
++          "oneOf": [
++            {
++              "type": "string",
++              "description": "Use any device.",
++              "enum": ["any"],
++              "default": "any"
++            },
++            {
++              "type": "object",
++              "description": "Dict of device hints to choose appropriate disk",
```

Comment:
@epim ?

Reply:
Personally I haven't had any direct need for using WWN, as we have fairly limited use of FC. However since it's part of the ironic root device hints we should include it here, as it will align better with the interface presented by Ironic.

Reply:
Yeah, I guess I don't really care if our downstream implementation has it, if users won't use it. But when we propose it upstream, we'll want it there for sure.